### ISSUE TITLE
More efficient Authentication ticket serialization

### ISSUE THREAD

**Issue by jods4** [aspnet/Security#589]
Serialized auth tickets can grow fairly large very quickly, even in basic scenarios. If you have moderately many roles, they all get serialized with the long claim type name and that's how your tickets becomes several Ko in size before you even know it.

In the common scenario where a ticket has a few of roles, today each one is adding more than 60 bytes just for its name (unencoded, actually more than that after encryption and encoding in base64). Having roles counted in tens results in cookies of size in thousands.

Previously the payload was compressed, but because of security concerns this was removed.

I noticed that in the beta-8 or rc-1 timeframe cookies chunking was added. That's a very good thing because it prevents failures caused by too large cookies. But it's not enough, because even if they don't fail, 10Ko cookies are not great for perf.

I know your guideline for very large tickets is to keep them on the server in session state. It's nice to have a solution to support scenarios where there is no way you could decently round-trip your state. But when you have a stateless server, you would like to keep it that way as much as possible as it is a lot simpler and safer.

With a more efficient serialization format, common auth. tickets could be a lot smaller and not require session state. I hope you'll pursue this option in addition to what has already been done.

Here's a simple suggestion: serialize claims grouped by claim type, so that you only have to serialize each claim type name only once. Something like this (pseudo-format):

```
http://schemas.xmlsoap.org/ws/2005/05/identity/claims/name
1
jods
http://schemas.microsoft.com/ws/2008/06/identity/claims/role
2
Admin
PowerUser
```

(Nitpicker corner: I know some claim types are multi-values, so it's more something like this: `Roles 2 1 Admin 1 PowerUser`, etc. But you get the idea.)

**Comment by brockallen** [comment]
> I know your guideline for very large tickets is to keep them on the server in session state.

Where did you see this guidance?

As to your issue -- I think the general consensus is that if you're putting that many claims into the cookie, then you might be designing your system incorrectly. Claims in the cookie should be identity data, not permissions data. You then use the identity data at runtime to locate additional data you at need for authorization. Additionally, you can always add claims transformation to your pipeline after the cookie middleware runs and make your user look as if it has all the roles in the cookie. And if your lookups are a concern, then do caching in the app layer.

**Comment by Tratcher** [comment]
@brockallen I assume by session he's referring to this: https://github.com/aspnet/Security/blob/dev/src/Microsoft.AspNet.Authentication.Cookies/CookieAuthenticationOptions.cs#L156-L160

**Comment by jods4** [comment]
@Tratcher yes, this is what I was referring to.

@brockallen 
At the extreme you can say that "identity data" should just be an id and I should cache everything and we're at the `SessionStore` solution. No need for more, right?

> I think the general consensus is that if you're putting that many claims into the cookie, then you might be designing your system incorrectly.

I'm doing what ASP.NET makes me do naturally. So maybe there's a design issue here?

- ASP.NET loves to handle permissions with roles. That's how it's been historically done and that's still very easy thanks to `User.IsInRole()`, `Authorize(Roles:="Admin")` and co.
- Roles are stored internally by ASP.NET as claims of a special type `http://schemas.microsoft.com/ws/2008/06/identity/claims/role`.
- When you use Cookie Authentication, by default all user claims are serialized and restored.

And just so you know, I'm not talking crazy stuff here. Serialized in an optimal way, my ticket could fit in a few 100s bytes. OK not possible for ASP.NET because of other constraints (security, general-purpose, etc.) but it sets a baseline. When you say _"that many claims"_ I'm saying in the 20-40 range for power users... is that crazy?

**Comment by brockallen** [comment]
Yea @Tratcher, if that's what he's referring to then it's misguided to think that means session state.

**Comment by brockallen** [comment]
> I'm doing what ASP.NET makes me do naturally. So maybe there's a design issue here?

Well, I write lots of ASP.NET code and it doesn't make me do that :) 

Perhaps you mean the templates? If so, then you may be onto something...

But as for roles -- look into the new policy-based and resource-based authorization systems. They've put good effort into helping you properly separate identity from permissions. 

ASP.NET and the [Authorize] attribute still supports roles for backwards compat. Also I think it'd just be an infinite uphill battle explaining to everyone that a role is just a claim. Not even Microsoft wants to tackle that education task.

**Comment by brockallen** [comment]
Also, as for why session state is so bad, here's some of it: http://brockallen.com/2012/04/07/think-twice-about-using-session-state/. Again, I think there's general consensus that the traditional session state is not an ideal mechanism for state management. 

The session interface @Tratcher linked is unrelated (but could be implemented in terms of session state, I suppose... I shudder at the though). Admittedly, I have not yet looked at what they've done in ASP.NET 5 for session state -- there is something for it, but my sense is that it's minimal (by design).

**Comment by jods4** [comment]
@brockallen `SessionState` can make you think about session, right? 
But indeed the default implementation is a `MemoryCache` which is not that bad, I agree.

> Perhaps you mean the templates? If so, then you may be onto something...

No, I mean the 3 bullets points I wrote above.

> ASP.NET and the [Authorize] attribute still supports roles for backwards compat.

Yes, and it's convenient. I know a role is just a basic "you have it or not" claim and that's fine by me. Is it an official MS message that they're for migrating old project and should be obsolete for new projects? I missed that message if it is.

> But as for roles -- look into the new policy-based and resource-based authorization systems.

I don't want to stray away from the issue I opened here so I'll skip the discussion about why policy-based is not a good fit for my current project and how I actually built some additional layers in my identity system and my roles are quite resources oriented and blah blah we don't care here.

Could you please explain to me into a little more details how using a resource-based authorization system would have avoided the issue at hand?

**Comment by brockallen** [comment]
Well done authorization has always been a challenge. Also, you have to keep in mind that the original design of the role-based system was really .NET 1.0's doing back in ~2002 (and no so much ASP.NET pushing you info any particular way) when active directory users and groups was the assumed security model. The world has changed since then -- apps need more than just roles to make authorization decisions. 

So... putting permissions in claims bloats the cookie (as you've identified), so that's a bad place (plus many permission checks also require a DB lookup). So in general it's best to put a minimal set of claims into the cookie and then as close to the point of authorization use that identity information to then go and load whatever other pertinent information is needed to make the authorization decision.

As for doing authorization (mainly via [Authorize] but also IsInRole checks), putting authorization logic into your app logic conflates the two (meaning not a clear separation of concerns). When that authorization logic needs to change you have lots of places to search & replace in the app code. Also, that logic ends up getting duplicated for UI as well (e.g. should the "edit" button be enabled/disabled). The idea with policy and resource based authorization (they're similar but different) is that the logic is centralized and decoupled from the rest of the app code. You can build unit tests to provide your authorization logic is working according to requirements. When the logic changes, you only have one place to change, re-test, and re-deploy. 

You of course need to invoke the authorization logic from the app logic (and ASP.NET 5 has an injectible authorization manager interface for this), but the point is that they're separated. Your app logic will no longer contain the "who's allowed" code and will be replaced with a single/simple "is the caller allowed" check. The logic for said check is then in the policy (the centralized code). It can accept custom classes (perhaps business objects) to assist with the authorization check and it can also make DB lookup calls (depending on your authorization rules/semantics).

So that's the poorly explained gist of it. Since 2009 with WIF and the concept of the claims authorization manager there has been Microsoft code/framework that encourages a separation of the app logic and authorization logic.

**Comment by leastprivilege** [comment]
That's not poorly explained at all. I wish the docs would express that as clearly as you did. Maybe Microsoft should copy that explanation

**Comment by jods4** [comment]
@brockallen Thanks for your time writing all of this but (no disrespect) you didn't really answer my question and are out of topic. This issue is about serialized size of auth tickets.

I know well enough the various models and as I mentioned briefly previously (because it's irrelevant to this ticket) I'm not using the "Roles" directly in the way you think. I called them "Roles" in the ticket because that's what ASP.NET calls them but in fact in my app they rather represent operations on resources and I have an extra layer between them and users (which app admins can change themselves). So as you can see I have already addressed the issues you list in your comment.

> putting permissions in claims bloats the cookie (as you've identified)

This is the whole point of this ticket and you didn't say why/how another model would sidestep this flaw. 

From an information theory standpoint, it seems to me that if you want fine-grained access rights you need to store that fine-grained data somewhere -- whatever the form. 
I think it comes down to: (a) serialize and round-trip (e.g. cookies); (b) fetch from your store (e.g. DB) on each request; and since this is slow you might do (b) but (c) cache the permissions server-side.

I still don't get how another authorization model would provide a new solution to that problem? If you have an answer to that question I would love to hear it.

Using a cache is OK and I have done it several times in the past. *It just doesn't feel right that the serialized format is so big and that a little more clever serialization could easily bring down the cookie size to a point of non-concern* (in my current project).

With that format, either you have little to no information _at all_ in the auth ticket, or you should configure `SessionStore` (but that's not going to be very intuitive to new coders, pit of success, etc.). If you don't pay attention and have moderate amounts of information your tickets will grow in the several Ko range _very quickly_.

As a final note, using cache is great but it is not 100% side-effects free (depending on your project requirements).

**Comment by brockallen** [comment]
> @brockallen Thanks for your time writing all of this but (no disrespect) you didn't really answer my question and are out of topic. This issue is about serialized size of auth tickets.

Nope, none taken. It's a good conversation. Half of what I'm saying is tacitly directed at Microsoft (but I'm sure they know that -- they're smart folks).

> This is the whole point of this ticket and you didn't say why/how another model would sidestep this flaw.

Well, I think I did -- the suggestion was to not put as many claims into the cookie/ticket. That's not an infinite resource, so don't treat it like one.

So I guess all the stuff I've posted on this thread wasn't directly addressing how to make the cookie smaller, but it's the suggestions I've been giving people for a while now to deal with the exact same issues we've had for the past 5+ years. That's why I was offering it. Sorry to derail your original issue.

> Using a cache is OK...

As for the cache -- the reason it's here now is probably because of how it was there in WIF. In WIF (and again, I'm assuming this is how Microsoft/ASP.NET 5 thinks of it) expects you issue a small cookie with just an identifier (e.g. GUID), then you do a cache on the server (reddis, DB, whatever -- but in-mem in server is prolly a bad idea). So yea, I think we're on same page WRT caching.

>  a little more clever serialization could easily bring down the cookie size to a point of non-concern

Maybe submit a PR to them with the improvements?

**Comment by jods4** [comment]
> the suggestion was to not put as many claims into the cookie/ticket.

But you have to store the information somewhere, right? Even if you switch your mental model, for the same amount of flexibility the same amount of information is required. And then you have to persist it somewhere.

So unless I miss the point, _"not put as many claims into the cookie"_ translates to "recompute permissions at each request" or maybe "fetch them from a cache".

Which are fine options, but for many scenarios, if the serialization format was not so bloated option "just put them in the cookie" (which is the built-in default behavior) would be just fine as well.
I think the fact that auth tickets were initially compressed is a sign that MS has been aware of this long before this ticket.

> Maybe submit a PR to them with the improvements?

If they are accepting PR for this I might, although _right now_ I really don't have time for that.

That said some design should be agreed upon beforehand, because there are trade-offs. For instance, what I proposed above is fine for my (and similar) cases only: when you have the same claim repeated lots of times. I don't know if it's worth thinking of other cases and what the alternative solutions may be.

I also note that if roles were a single multi-valued claim, it would already work like the serialization I suggested. And it would make more sense I think (to have a single claim _Roles = A,B,C_).
I guess that might be a compatibility issue / breaking change?

**Comment by Zoxive** [comment]
Just hit this the other day, our cookie sizes balloon from many security roles. (We use SSO so users can get many many security roles)
Its a bit difficult to configure compression yourself with the current setup and feels like it should either be on by default, or an easy opt in via ``CookieAuthenticationOptions``.

The main problem is that when configuring cookie options you dont have an option to set the ``IDataSerializer<AuthenticationTicket>`` directly, instead you have to create your serializer, and get access to ``IDataProtectionProvider`` which is in dependancy injection, so you have to write another class to register a ``IConfigureNamedOptions<CookieAuthenticationOptions>``

Heres the changes a user has to do:
https://github.com/aspnet/Security/compare/dev...Zoxive:CompressCookies

**Comment by Tratcher** [comment]
@Zoxive do not compress encrypted content. See https://en.wikipedia.org/wiki/CRIME

**Comment by Zoxive** [comment]
@Tratcher To be clear its encrypting compressed content.

**Comment by Eilon** [comment]
Closing because no further action is planned on this issue.
